### ISSUE TITLE
[Bug] Device of `DDPTensorizedDataset` is always cpu

### ISSUE THREAD

**Issue by yaox12** [dmlc/dgl#3719]
## 🐛 Bug

https://github.com/dmlc/dgl/blob/fa343873a87dadbc64f08f3b247f8be7fc9f94ff/python/dgl/dataloading/dataloader.py#L176-L183

Shared memory here can only be on CPU, which results in the output of `DDPTensorizedDataset` is always on CPU.

Actually, I think what should be shared here are indices of `_tensor_dataset`. The first argument passed in is `train_ids` created in the training scripts and there is no need to share it. 

@BarclayII 

## Environment

 - DGL Version (e.g., 1.0): TOT

**Comment by BarclayII** [comment]
What is the difference between sharing the indices of `_tensor_dataset` and `_tensor_dataset` itself if `_tensor_dataset` is a Tensor?  And we do need to share it somehow because we need to shuffle it every epoch.

**Comment by yaox12** [comment]
> What is the difference between sharing the indices of `_tensor_dataset` and `_tensor_dataset` itself if `_tensor_dataset` is a Tensor? 

Because `_tensor_dataset` itself can be a GPU tensor.

> And we do need to share it somehow because we need to shuffle it every epoch.

For a typical use case of ddp dataloader,
https://github.com/dmlc/dgl/blob/fa343873a87dadbc64f08f3b247f8be7fc9f94ff/examples/pytorch/graphsage/train_sampling_multi_gpu.py#L85-L94
`train_nid` here is assumed to be the same in each process.

**Comment by BarclayII** [comment]
> Because `_tensor_dataset` itself can be a GPU tensor.

Would it be the same GPU tensor across all trainer processes?  I remember I had problems with PyTorch when it comes to multiple trainer processes running on the same GPU.  If it's OK to access the same GPU tensor across different trainer processes then I think we can change the shared memory array to share indices.

**Comment by yaox12** [comment]
> Would it be the same GPU tensor across all trainer processes?

I think it's duplicated in each process and on each GPU.
